### DISCUSSION TITLE
Reconnection to redis cluster

### DISCUSSION THREAD

**Discussion by Krzysztof-Lempicki** [xetorthio/jedis#3081] [Q&A]
I am using spring boot and jedis 2.7.1

# Problem:
When my app is started before redis cluster aftere few retries 
I receive:

`CLUSTERDOWN Hash slot not served`

or 

`No reachable node in cluster`

If cluster is up jedis still shows me this error.

# Question:
How to enable jedis to try reconnect to cluster until it is present?

**Comment by sazzad16** [comment]
@Krzysztof-Lempicki Your "cluster node" may be up but your "cluster" is not up (and stable). Jedis can't automatically fix this.

Please try to Google the error and see which is your exact problem and what is the solution. Here is the top result https://stackoverflow.com/questions/36125071/redis-cluster-master-slave-not-able-to-add-key

**Comment by Krzysztof-Lempicki** [comment]
My situations looks like this:

1.  Spring boot app started - `redis` cluster was not up yet.
2.  `Jedis` is trying to connect five times.
3.  Redis cluster is up after five retires from point 2
4.  Any request to redis with `Jedis` gives me `No reachable node in cluster`
5. I am restarting my SpringBoot app and everything is working fine.

What is the safest way to do this without restarting app?
I can use aspect but I don't know what this aspect need to do to 
reconnect?

**Comment by Krzysztof-Lempicki** [comment]
I created custom connection handler: 
```
public class JedisAutoReconnectConnectionHandler extends JedisSlotBasedConnectionHandler {

    public JedisAutoReconnectConnectionHandler(Set<HostAndPort> nodes,
                                               GenericObjectPoolConfig<Jedis> poolConfig,
                                               int connectionTimeout,
                                               int soTimeout,
                                               String password,
                                               String clientName) {
        super(nodes, poolConfig, connectionTimeout, soTimeout, password, clientName);
    }

    @Override
    public Jedis getConnection() {
        try {
            return super.getConnection();
        } catch (JedisNoReachableClusterNodeException exception) {
            this.renewSlotCache();
            throw exception;
        }
    }

    @Override
    public Jedis getConnectionFromSlot(int slot) { // FIXME no thread safe
        try {
            return super.getConnectionFromSlot(slot);
        } catch (JedisNoReachableClusterNodeException exception) {
            this.renewSlotCache();
            throw exception;
        }
    }
}

```
and added it to my cluster but this changes nothing. Only app restart helps.

**Comment by sazzad16** [comment]
[This change](https://github.com/redis/jedis/pull/2643) seemingly resolves your issue. But it is included in Jedis 4.2.0 and Spring has not officially published Jedis 4 support yet. You can use Spring 3.0 snapshot which supports Jedis 4.2.x.

**Comment by Krzysztof-Lempicki** [comment]
I tried `version 4.2`  and it does not solve this problem.

`ClusterConnectionProvider`  has method:

```
  private void initializeSlotsCache(Set<HostAndPort> startNodes, JedisClientConfig clientConfig) {
    ArrayList<HostAndPort> startNodeList = new ArrayList<>(startNodes);
    Collections.shuffle(startNodeList);

    for (HostAndPort hostAndPort : startNodeList) {
      try (Connection jedis = new Connection(hostAndPort, clientConfig)) {
        cache.discoverClusterNodesAndSlots(jedis);
        return;
      } catch (JedisConnectionException e) {
        // try next nodes
      }
    }

    throw new JedisClusterOperationException("Could not initialize cluster slots cache.");
  }
```

To make it short: if no redis cluster is present when app start throw exception

**Reply by sazzad16** [reply]
You're right. I forgot about this change for a moment.

Your requirement is different from all the feature requests and issues we ever have regarding cluster mode.
